Source/Scene/ShadowMap.js

```diff
@@ -0,0 +1,397 @@
+            });
+
+            // Create the light
+            shadowMap._debugLight = new Primitive({
+                geometryInstances : new GeometryInstance({
+                    geometry : new SphereOutlineGeometry({
+                        radius : 1.0
+                    }),
+                    modelMatrix : modelMatrix,
+                    attributes : {
+                        color : ColorGeometryInstanceAttribute.fromColor(Color.BLUE)
+                    }
+                }),
+                appearance : new PerInstanceColorAppearance({
+                    translucent : false,
+                    flat : true
+                }),
+                asynchronous : false
+            });
+
+            //// Create the camera frustum
+            //var nearPlane = frameState.camera.frustum.near;
+            //var farPlane = frameState.camera.frustum.far;
+            //shadowMap._debugCameraFrustum = new Primitive({
+            //    geometryInstances : new GeometryInstance({
+            //        geometry : new BoxOutlineGeometry({
+            //            minimum : new Cartesian3(-0.0, -0.0, -nearPlane),
+            //            maximum : new Cartesian3(0.0, 0.0, -farPlane)
+            //        }),
+            //        attributes : {
+            //            color : ColorGeometryInstanceAttribute.fromColor(Color.RED)
+            //        }
+            //    }),
+            //    appearance : new PerInstanceColorAppearance({
+            //        translucent : false,
+            //        flat : true
+            //    }),
+            //    asynchronous : false
+            //});
+        }
+
+        //if (!shadowMap.debugFreezeFrame) {
+        //
+        //    //var nearPlane = frameState.camera.frustum.near;
+        //    //var point = Matrix4.multiplyByPoint(frameState.context.uniformState.inverseViewProjection, new Cartesian3(0, 0, -nearPlane), new Cartesian3());
+        //    //console.log(point.x, point.y, point.z);
+        //    //console.log(frameState.camera.positionWC.x, frameState.camera.positionWC.y, frameState.camera.positionWC.z);
+        //    //
+        //    //// TODO : possibly get this from the camera instead
+        //    //var cameraMatrix = frameState.context.uniformState.inverseViewProjection;
+        //    //shadowMap._debugCameraFrustum.modelMatrix = cameraMatrix;
+        //}
+
+        shadowMap._debugLightFrustum.update(frameState);
+        shadowMap._debugLight.update(frameState);
+        //shadowMap._debugCameraFrustum.update(frameState);
+
+        updateDebugShadowViewCommand(shadowMap, frameState);
+    }
+
+    // Converts from NDC space to texture space
+    var scaleBiasMatrix = new Matrix4(0.5, 0.0, 0.0, 0.5, 0.0, 0.5, 0.0, 0.5, 0.0, 0.0, 0.5, 0.5, 0.0, 0.0, 0.0, 1.0);
+    var scratchMatrix = new Matrix4();
+
+    ShadowMap.prototype.update = function(frameState) {
+        var enabled = frameState.shadowsEnabled = this.enabled;
```

Comment:
The hidden logic and initialization here and then again in `executeShadowMapCommands` is hard to follow.

Instead, handle all the `enabled` logic in `executeViewportCommands`.